Set Mautic configuration param via .env file

alexander-nitsche-governikus opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🟢 <!-- Use emojis to indicate positive (green) or negative (red) for each item in the table. -->
| New feature/enhancement? (use the a.x branch)      | 🔴
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🟢 <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      |  <!-- required for new features -->
| Related developer documentation PR URL |  <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #14500  <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

## Description

Override Mautic configuration param via system and DotEnv environment variable "MAUTIC_CONFIG_PARAMETERS".
Resolve Mautic configuration param "getenv(..)" function via system and DotEnv environment variables.

For example set the context-specific database configuration via .env.local:

    MAUTIC_DB_HOST=..
    MAUTIC_DB_NAME=..
    MAUTIC_DB_USER=..
    MAUTIC_DB_PASSWORD=..
    MAUTIC_DB_PORT=..

and resolve in local.php via:

    'db_host' => 'getenv(MAUTIC_DB_HOST)',
    'db_name' => 'getenv(MAUTIC_DB_NAME)',
    'db_user' => 'getenv(MAUTIC_DB_USER)',
    'db_password' => 'getenv(MAUTIC_DB_PASSWORD)',
    'db_port' => 'getenv(MAUTIC_DB_PORT)',

or override the same parameters via .env.local:

    MAUTIC_CONFIG_PARAMETERS='{"db_host":"..","db_name":"..","db_user":"..","db_password":"..","db_port":".."}'

Previously this was only supported via system environment variables. 

Furthermore "getenv()" parameter values get resolved for the final parameter bags as well as for the final parameter environment variables.

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->
<!-- Remove HTML comment markup below to use the table for screenshots when relevant. -->
<!--
| Before                                 | After
| -------------------------------------- | ---
|                                        | 
-->


---
### 📋 Steps to test this PR:

<!--
This part is crucial. Take the time to write very clear, annotated and step by step test instructions, because testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Confirm that the Mautic frontend is rendering fine in your browser.
3. In the .env.local file, add "MAUTIC_DB_HOST=&lt;db-host&gt;" with &lt;db-host&gt; being a non-existing hostname.
4. In the local.php file, set "'db_host' => 'getenv(MAUTIC_DB_HOST)'".
5. Confirm that the Mautic frontend is throwing an error in your browser.
6. Revert and confirm that the Mautic frontend is rendering fine in your browser again.
7. In the system environment, e.g. via the docker-compose file, add "MAUTIC_DB_HOST=&lt;db-host&gt;" with &lt;db-host&gt; being a non-existing hostname. Restart the system.
8. Confirm that the Mautic frontend is throwing an error in your browser.
9. Revert and confirm that the Mautic frontend is rendering fine in your browser again.

<!--
If you have any deprecations and backwards compatibility breaks, list them here along with the new alternative.
-->

alexander-nitsche-governikus committed:
Set Mautic configuration param via .env file

Override Mautic configuration param <configuration_param> with
<MAUTIC_CONFIGURATION_PARAM> in .env files.

For example set the context-specific database configuration via
.env.local:

    MAUTIC_DB_HOST=..
    MAUTIC_DB_NAME=..
    MAUTIC_DB_USER=..
    MAUTIC_DB_PASSWORD=..
    MAUTIC_DB_PORT=..

Previously this was only supported via system environment variables.

kuzmany commented:
We can currently achieve this using the JSON array data stored in MAUTIC_CONFIG_PARAMETERS.
For more information, you can refer to this link: 

https://github.com/mautic/mautic/blob/1f1418da6f82697d44a2af7d47e0709a1e9bd2cf/app/bundles/CoreBundle/Loader/ParameterLoader.php#L175

Does this not meet your needs?

alexander-nitsche-governikus commented:
hi @kuzmany , unforetunately, the parameteres of .env files are not loaded into the system environment, but only into $_ENV and $_SERVER, so they cannot be retrieved via getenv().

escopecz commented:
Please check the failing tests

alexander-nitsche-governikus committed:
Revert "Set Mautic configuration param via .env file"

This reverts commit 2389d0faa98448d108c24b557d6be8b24eff33f3.

alexander-nitsche-governikus committed:
Fix resolving configuration params via environment variables

With this patch, Mautic resolves the configuration parameters in this
order:

1. Parameters via system environment variable "MAUTIC_CONFIG_PARAMETERS"
2. Parameters via DotEnv environment variable "MAUTIC_CONFIG_PARAMETERS"
4. Parameters in local parameters file "parameters_local.php"
3. Parameters in local config file "local.php"
5. Default parameters

The parameter value "getenv(...)" gets parsed early and considers the
environment variables of system and DotEnv.

alexander-nitsche-governikus commented:
Hi Zdeno, hi John, i reworked the patch as i recognized that the parameter loader processes in this order:

1) Grabbing configuration parameters.
2) Loading configuration parameters into environment.

and i was focused on step (2) only. Now i have moved to step (1) which subsequently has consequences for step (2). I expanded the parsing of environment variables in (1) and moved this logic into a new class \Mautic\CoreBundle\Loader\Parameters\EnvVarParameters - heavily inspired by \Mautic\CoreBundle\Loader\EnvVars\ConfigEnvVars et al.. One could consider to add additional parameter loader for the remaining inline blocks of \Mautic\CoreBundle\Loader\ParameterLoader::loadLocalParameters().

alexander-nitsche-governikus committed:
Fix PHPStan findings

alexander-nitsche-governikus committed:
Fix tests

alexander-nitsche-governikus commented:
PHPStan should be satisfied. PHPunit should be satisfied too. The only tests that do not pass locally are SAML tests and Installation tests. I think these failures are related to my local setup rather than to this patch. Hope for green lights in a fresh Github workflow run.

alexander-nitsche-governikus committed:
Fix PHPStan findings

alexander-nitsche-governikus commented:
Hi Martin, thanks for the hint, i have rebased the branch.

matbcvo commented:
Is this PR ready for review?

alexander-nitsche-governikus commented:
Yes. I am already using it as a patch in my current mautic v5 instance and so far it works as expected.

RCheesley commented:
I'm bumping this from the milestone as we're still pending testing/code review etc.

Bastian2718 commented:
Tested Step 1 to 6 and works very well but how can I test the other steps in gitpod? Or do I need to test this locally?

alexander-nitsche-governikus commented:
Hi Bastian, i tested in locally in my custom Docker Compose environment. I have no experience with Gitpod, sorry. Greetings, Alex

RCheesley commented:
I've asked the Docker Maintainers working group to take a look at this PR so you can get a few more 👀 on it.

henmohr reviewed (approved):
I think it's ok but i'm no programmer xD

cibero42 reviewed (approved):
Well, I'm no PHP specialist, but the proposition makes total sense for me and might help with some Docker related issues.

matbcvo commented:
Please rebase your PR onto the 6.0 branch.
